File: rust/notify-agent-windows/src/tray.rs
//! System tray icon for the Windows agent (design: ported from the C# TrayApplicationContext).
//! Owns a hidden window whose sole purpose is to host a `Shell_NotifyIconW` tray icon and pump
//! the Win32 messages it needs (tray click callback, context menu). Runs on the thread that
//! calls `run_message_loop`, which blocks for the app's whole lifetime — the tokio runtime and
//! `AgentHost` live on a separate thread (see `main.rs`), since a Win32 message pump and a
//! blocking `block_on` can't share one thread.

use std::sync::mpsc;
use std::time::Duration;

use windows::core::{HSTRING, PCWSTR};
use windows::Win32::Foundation::{HINSTANCE, HWND, LPARAM, LRESULT, WPARAM};
use windows::Win32::System::LibraryLoader::GetModuleHandleW;
use windows::Win32::UI::Shell::{
    Shell_NotifyIconW, NIF_ICON, NIF_MESSAGE, NIF_TIP, NIM_ADD, NIM_DELETE, NIM_MODIFY,
    NOTIFYICONDATAW,
};
use windows::Win32::UI::WindowsAndMessaging::{
    AppendMenuW, CreatePopupMenu, CreateWindowExW, DefWindowProcW, DispatchMessageW, GetCursorPos,
    GetMessageW, GetWindowLongPtrW, LoadCursorW, LoadIconW, PostMessageW, RegisterClassExW,
    SetForegroundWindow, SetWindowLongPtrW, TrackPopupMenu, TranslateMessage, CREATESTRUCTW,
    GWLP_USERDATA, HMENU, IDC_ARROW, MF_DISABLED, MF_SEPARATOR, MF_STRING, MSG,
    TPM_BOTTOMALIGN, TPM_LEFTALIGN, TPM_RIGHTBUTTON, WM_APP, WM_COMMAND, WM_LBUTTONUP, WM_NCCREATE,
    WM_NULL, WM_RBUTTONUP, WNDCLASSEXW, WS_OVERLAPPED,
};

const CLASS_NAME: &str = "NotifyAgentRustTrayWindow";
const BASE_TOOLTIP: &str = "Desktop Notification Agent";
const TRAY_UID: u32 = 1;
const WM_TRAYICON: u32 = WM_APP + 1;
const ID_MENU_VERSION: usize = 1001;
const ID_MENU_CLOSE: usize = 1002;
const CLOSE_TIMEOUT: Duration = Duration::from_secs(5);
/// Resource id of the embedded icon (see `icon.rc` / `assets/app.ico`, compiled in via `build.rs`).
const APP_ICON_ID: u16 = 1;

/// `Shell_NotifyIconW` is documented safe to call from any thread, so this small handle (just
/// the window handle) is shared with the agent thread to flag a startup failure directly,
/// without hopping back onto the window thread the way the WinForms `NotifyIcon` needs to.
#[derive(Clone, Copy)]
pub struct TrayHandle(HWND);

unsafe impl Send for TrayHandle {}

impl TrayHandle {
    pub fn set_start_failed(&self) {
        let mut data = NOTIFYICONDATAW {
            cbSize: std::mem::size_of::<NOTIFYICONDATAW>() as u32,
            hWnd: self.0,
            uID: TRAY_UID,
            uFlags: NIF_TIP,
            ..Default::default()
        };
        set_tip(
            &mut data,
            &format!("{BASE_TOOLTIP} (agent failed to start)"),
        );
        unsafe {
            let _ = Shell_NotifyIconW(NIM_MODIFY, &data);
        }
    }
}

struct TrayState {
    menu: HMENU,
    close_tx: tokio::sync::mpsc::UnboundedSender<()>,
    done_rx: std::cell::RefCell<Option<mpsc::Receiver<()>>>,
}

fn set_tip(data: &mut NOTIFYICONDATAW, text: &str) {
    let wide: Vec<u16> = text.encode_utf16().take(data.szTip.len() - 1).collect();
    data.szTip[..wide.len()].copy_from_slice(&wide);
    data.szTip[wide.len()] = 0;
}

/// Creates the hidden tray window and shows the icon immediately (before the agent has
/// finished starting, matching the C# design's "icon appears without waiting on NATS connect").
/// `close_tx` is signaled (from the window thread) when the user clicks Close; `done_rx` is
/// consumed by the Close handler's watcher thread, which waits up to `CLOSE_TIMEOUT` for the
/// agent thread to report shutdown finished, then hard-exits the process regardless.
pub fn create(
    close_tx: tokio::sync::mpsc::UnboundedSender<()>,
    done_rx: mpsc::Receiver<()>,
) -> anyhow::Result<TrayHandle> {
    unsafe {
        let hinstance: HINSTANCE = GetModuleHandleW(PCWSTR::null())?.into();
        let class_name = HSTRING::from(CLASS_NAME);

        let wc = WNDCLASSEXW {
            cbSize: std::mem::size_of::<WNDCLASSEXW>() as u32,
            lpfnWndProc: Some(wndproc),
            hInstance: hinstance,
            hCursor: LoadCursorW(None, IDC_ARROW)?,
            lpszClassName: PCWSTR(class_name.as_ptr()),
            ..Default::default()
        };
        if RegisterClassExW(&wc) == 0 {
            anyhow::bail!(
                "RegisterClassExW failed: {:?}",
                windows::core::Error::from_win32()
            );
        }

        let menu = CreatePopupMenu()?;
        AppendMenuW(
            menu,
            MF_STRING | MF_DISABLED,
            ID_MENU_VERSION,
            &HSTRING::from(format!("Version {}", env!("CARGO_PKG_VERSION"))),
        )?;
        AppendMenuW(menu, MF_SEPARATOR, 0, PCWSTR::null())?;
        AppendMenuW(menu, MF_STRING, ID_MENU_CLOSE, &HSTRING::from("Close"))?;

        let state = Box::into_raw(Box::new(TrayState {
            menu,
            close_tx,
            done_rx: std::cell::RefCell::new(Some(done_rx)),
        }));

        let hwnd = CreateWindowExW(
            Default::default(),
            &class_name,
            &HSTRING::from(CLASS_NAME),
            WS_OVERLAPPED,
            0,
            0,
            0,
            0,
            None,
            None,
            hinstance,
            Some(state as *const std::ffi::c_void),
        )?;

        let mut data = NOTIFYICONDATAW {
            cbSize: std::mem::size_of::<NOTIFYICONDATAW>() as u32,
            hWnd: hwnd,
            uID: TRAY_UID,
            uFlags: NIF_MESSAGE | NIF_ICON | NIF_TIP,
            uCallbackMessage: WM_TRAYICON,
            hIcon: LoadIconW(hinstance, PCWSTR(APP_ICON_ID as _))?,
            ..Default::default()
        };
        set_tip(&mut data, BASE_TOOLTIP);
        Shell_NotifyIconW(NIM_ADD, &data).ok()?;

        Ok(TrayHandle(hwnd))
    }
}

/// Blocks the calling thread pumping Win32 messages for the app's whole lifetime. Only
/// returns if the message loop is torn down (WM_QUIT) or GetMessageW errors — in practice
/// the process exits via `std::process::exit` from the Close watcher thread before that
/// ever happens.
pub fn run_message_loop() {
    let mut msg = MSG::default();
    unsafe {
        loop {
            // GetMessageW returns 0 on WM_QUIT, -1 on error, nonzero otherwise — a plain
            // bool conversion would treat -1 as "message received" and dispatch `msg`
            // uninitialized/stale.
            if GetMessageW(&mut msg, None, 0, 0).0 <= 0 {
                break;
            }
            let _ = TranslateMessage(&msg);
            DispatchMessageW(&msg);
        }
    }
}

fn state_from_hwnd(hwnd: HWND) -> Option<&'static TrayState> {
    unsafe {
        let ptr = GetWindowLongPtrW(hwnd, GWLP_USERDATA) as *const TrayState;
        ptr.as_ref()
    }
}

unsafe extern "system" fn wndproc(hwnd: HWND, msg: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
    if msg == WM_NCCREATE {
        let create_struct = &*(lparam.0 as *const CREATESTRUCTW);
        SetWindowLongPtrW(hwnd, GWLP_USERDATA, create_struct.lpCreateParams as isize);
        return DefWindowProcW(hwnd, msg, wparam, lparam);
    }

    let Some(state) = state_from_hwnd(hwnd) else {
        return DefWindowProcW(hwnd, msg, wparam, lparam);
    };

    match msg {
        WM_TRAYICON if matches!(lparam.0 as u32, WM_RBUTTONUP | WM_LBUTTONUP) => {
            let mut pt = Default::default();
            let _ = GetCursorPos(&mut pt);
            let _ = SetForegroundWindow(hwnd); // so the menu dismisses on click-away
            let _ = TrackPopupMenu(
                state.menu,
                TPM_RIGHTBUTTON | TPM_BOTTOMALIGN | TPM_LEFTALIGN,
                pt.x,
                pt.y,
                0,
                hwnd,
                None,
            );
            // Required after TrackPopupMenu per MSDN, else the menu can fail to close.
            let _ = PostMessageW(hwnd, WM_NULL, WPARAM(0), LPARAM(0));
            LRESULT(0)
        }
        // WM_COMMAND packs the notification code in HIWORD(wParam); it's 0 for menu clicks,
        // but mask it explicitly rather than relying on that always being the case.
        WM_COMMAND if (wparam.0 & 0xFFFF) == ID_MENU_CLOSE => {
            let data = NOTIFYICONDATAW {
                cbSize: std::mem::size_of::<NOTIFYICONDATAW>() as u32,
                hWnd: hwnd,
                uID: TRAY_UID,
                ..Default::default()
            };
            let _ = Shell_NotifyIconW(NIM_DELETE, &data);
            let _ = state.close_tx.send(());
            if let Some(done_rx) = state.done_rx.borrow_mut().take() {
                std::thread::spawn(move || {
                    let _ = done_rx.recv_timeout(CLOSE_TIMEOUT);
                    std::process::exit(0);
                });
            }
            LRESULT(0)
        }
        _ => DefWindowProcW(hwnd, msg, wparam, lparam),
    }
}

File: golang/cmd/notify-agent-windows/tray.go
//go:build windows

package main

import (
	"context"
	_ "embed"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"sync/atomic"
	"time"

	"github.com/getlantern/systray"

	"github.com/YuShuanHsieh/os-notification/golang/internal/host"
	"github.com/YuShuanHsieh/os-notification/golang/internal/metrics"
	"github.com/YuShuanHsieh/os-notification/golang/internal/natsauth"
)

// version is the running agent's version string, shown (disabled/
// non-clickable) in the tray context menu as "Version 0.1.0" — matching the
// C# TrayApplicationContext's "Version {VersionInfo.Current}" and the Rust
// tray's "Version {CARGO_PKG_VERSION}", both currently "0.1.0".
const version = "0.1.0"

const baseTooltip = "Desktop Notification Agent"

// closeTimeout bounds how long the Close handler waits for a graceful
// AgentHost.Shutdown before force-exiting the process regardless — the same
// 5-second bound used by tray.rs's CLOSE_TIMEOUT and
// TrayApplicationContext.CloseTimeout.
const closeTimeout = 5 * time.Second

//go:embed assets/app.ico
var appIcon []byte

// trayApp owns the tray icon/menu and the AgentHost's lifecycle.
//
// Lifecycle: systray.Run(onReady, onExit) synchronously registers the
// window and adds the tray icon (NIM_ADD, with a default placeholder icon)
// before onReady ever runs, and onReady itself always runs on its own
// goroutine, separate from the native message-pump thread. So the icon is
// visible immediately at launch, and startAgent's NATS/host.Start work
// (which may block or fail) never blocks the tray icon or its message pump
// — matching the C# TrayApplicationContext / Rust tray.rs design of
// "icon appears without waiting on NATS connect".
type trayApp struct {
	h atomic.Pointer[host.Host]

	// settings is the Feature-2 settings-file value loaded once at process
	// startup (main.go), before systray.Run -- it never changes for the
	// life of the process, so no synchronization is needed to read it from
	// startAgent's goroutine.
	settings Settings

	// metrics is the AgentMetrics constructed once at process startup
	// (main.go's InitMetrics call), before systray.Run -- like settings, it
	// never changes for the life of the process, so no synchronization is
	// needed to read it from startAgent's goroutine. Nil-safe: a zero-value
	// trayApp (as tests might construct) has a nil metrics field, but
	// host.Start treats a nil AgentMetrics the same as
	// metrics.NullAgentMetrics{}.
	metrics metrics.AgentMetrics
}

func (a *trayApp) onReady() {
	systray.SetIcon(appIcon)
	systray.SetTooltip(baseTooltip)
	slog.Info("tray icon shown", "version", version)

	versionItem := systray.AddMenuItem(fmt.Sprintf("Version %s", version), "")
	versionItem.Disable()
	systray.AddSeparator()
	closeItem := systray.AddMenuItem("Close", "Shut down the agent and exit")

	go a.watchClose(closeItem)
	go a.startAgent()
}

// onExit is systray's exit callback, invoked from systray.Quit() (called by
// watchClose). There is nothing left to do here: watchClose already owns
// the bounded graceful shutdown and the process's forced exit.
func (a *trayApp) onExit() {}

// startAgent starts the AgentHost. On failure, the tray icon and its Close
// item remain present and usable — Close still works with no host to shut
// down (mirrors the C# design's failure path) — but the tooltip is updated
// to flag the failure, per the "(agent failed to start)" requirement.
func (a *trayApp) startAgent() {
	opts := ResolveHostOptions(os.Getenv, host.OptionsFromEnv(), a.settings)

	var authProvider natsauth.Provider
	if credsPath := ResolveCredsFile(os.Getenv, a.settings); strings.TrimSpace(credsPath) != "" {
		authProvider = natsauth.CredsFileAuth{Path: credsPath}
	}

	renderer := NewWindowsRenderer(defaultImageCacheDir())

	// WindowsUsernameIdentity (identity_windows.go) derives identity from
	// the current Windows account name via GetUserNameW -- a deliberate,
	// documented exception to "the OS account name is never used as
	// identity" scoped to the Windows heads (see identity.go's package
	// doc). This Go port has no AAD/MSAL/device-code sign-in at all, so it
	// is used unconditionally rather than only as an AAD fallback.
	idp := WindowsUsernameIdentity{Getenv: os.Getenv, Settings: a.settings}
	h, err := host.Start(context.Background(), opts, idp, renderer, authProvider, a.metrics)
	if err != nil {
		// slog.Error alone is sufficient here: main.go installs a text
		// handler that writes to os.Stderr, so a separate raw
		// fmt.Fprintf(os.Stderr, ...) would just double-report this same
		// failure, once structured and once raw.
		slog.Error("agent failed to start", "error", err)
		systray.SetTooltip(fmt.Sprintf("%s (agent failed to start)", baseTooltip))
		return
	}
	a.h.Store(h)
}

// watchClose waits for the Close menu item to be clicked, then attempts a
// bounded graceful AgentHost.Shutdown (up to closeTimeout), then hides the
// icon and force-exits the process — mirroring tray.rs's
// WM_COMMAND/ID_MENU_CLOSE handler and TrayApplicationContext.OnCloseClickedAsync.
func (a *trayApp) watchClose(closeItem *systray.MenuItem) {
	<-closeItem.ClickedCh
	slog.Info("close clicked, shutting down")

	// Run the bounded graceful shutdown to completion (or timeout) BEFORE
	// calling systray.Quit(). systray.Quit() causes the blocking
	// systray.Run() call in main()'s run() to return, which immediately
	// hits os.Exit(0) there — if Quit() fired first, that race could exit
	// the process before Shutdown ever got a chance to run, defeating the
	// whole point of a graceful, bounded Close.
	done := make(chan struct{})
	go func() {
		defer close(done)
		if h := a.h.Load(); h != nil {
			ctx, cancel := context.WithTimeout(context.Background(), closeTimeout)
			defer cancel()
			_ = h.Shutdown(ctx)
		}
	}()

	select {
	case <-done:
	case <-time.After(closeTimeout):
	}

	systray.Quit() // hides the icon and tears down the message loop
	os.Exit(0)
}
